Hide non alphabetic tokens

gfontorbe opened the pull request:
This is a proposal for addressing #268 

It adds a new boolean entry in `langium-config.ts` named "showNonAlphabeticKeywords". It allows for the completion provider to skip keywords that are composed of a single non-alphabetic character. If not present in the config file, it defaults to `false`

With "showNonAlphabeticKeywords" set to `true`:
<img width="661" alt="Screenshot 2022-09-23 at 14 47 47" src="https://user-images.githubusercontent.com/44747557/191964740-ba92a250-1054-4d3a-b547-a13abedac13e.png">

With "showNonAlphabeticKeywords" set to `false`:
<img width="633" alt="Screenshot 2022-09-23 at 14 48 23" src="https://user-images.githubusercontent.com/44747557/191964805-23e043c9-9204-4042-800e-1474bb9bec70.png">

One thing I didn't manage to achieve is to add support for this new entry in `langium-config.ts`:
<img width="569" alt="Screenshot 2022-09-23 at 14 48 50" src="https://user-images.githubusercontent.com/44747557/191964897-8ded5355-8827-425e-a5bd-755f2d42500f.png">
The json validation shows an error but the module is generated correctly.


PS: I didn't adapt the completion tests yet. Will do in the same PR if the approach is accepted

Lotes commented on packages/langium/src/lsp/completion/completion-provider.ts:
```diff
@@ -239,6 +242,9 @@ export class DefaultCompletionProvider implements CompletionProvider {
     }
 
     protected completionForKeyword(keyword: ast.Keyword, context: AstNode | undefined, acceptor: CompletionAcceptor): MaybePromise<void> {
+        if (keyword.value.match(/^[\W]/) && keyword.value.length === 1 && !this.languageMetadata.showNonAlphabeticKeywords) {
```
Hard to read... what is happening?

Lotes replied:
Starts with a non word character of length 1?

gfontorbe replied:
This checks if the the keyword start with a non-alphabetic token and is a single character.
I don't think hiding keywords such as `_keyword` from completion would be correct.

It also checks if the option `showNonAlphabeticKeywords ` is set to false.

For better readability, I guess we could first check for the option in an other if. What do you think?

Lotes replied:
Yes, maybe a nested IF or a predicate function for the other two condition...
```js
function isNonAlphabeticToken() {
  return keyword.value.match(/^[\W]/) && keyword.value.length === 1;
}
```

Lotes commented on packages/langium/src/lsp/completion/completion-provider.ts:
```diff
@@ -239,6 +242,9 @@ export class DefaultCompletionProvider implements CompletionProvider {
     }
 
     protected completionForKeyword(keyword: ast.Keyword, context: AstNode | undefined, acceptor: CompletionAcceptor): MaybePromise<void> {
+        if (keyword.value.match(/^[\W]/) && keyword.value.length === 1 && !this.languageMetadata.showNonAlphabeticKeywords) {
```
The name `showNonAlphabeticKeywords` makes it a bit hart to think about it:
1. `show` -> ok
2. `non-alphabetic-keywords`
   1. `keyword` makes sense
   2. `alphabethic`, mhhh ok like ABC
   3. `non` = negating the term before
3. now try to read `!non-alphabetic-keywords` , what is it?

Proposal: `showOnlyAlphabeticKeywords`... ok, when NOTTING it is a bit the same. ^^*... so even I am not sure about this solution...

gfontorbe replied:
That would remove the need for negation in the if statement

spoenemann reviewed (commented):
Proposal: Omit the new configuration property and hard-code the change to `completionForKeyword` as the new default behavior. If someone wants to change this, they can simply override the method in a subclass of the CompletionProvider service.

Advantages:
 - Keeps the configuration format simpler
 - More flexible: users can change the regular expression to something else if they want to fine-tune which keywords are shown

gfontorbe commented:
I went with @spoenemann approach and hardcoded the new behavior in the completion provider

gfontorbe commented:
I changed the regex to check if the keyword does not contain any word character, this way we also filter out keywords like `>=` `&&` ...

I also modified 3 tests that were expecting keywords with no word characters inside.

spoenemann commented on packages/langium/src/lsp/completion/completion-provider.ts:
```diff
@@ -239,6 +239,10 @@ export class DefaultCompletionProvider implements CompletionProvider {
     protected completionForKeyword(keyword: ast.Keyword, context: AstNode | undefined, acceptor: CompletionAcceptor): MaybePromise<void> {
+        // Filter out keywords that do not contain any word character
+        if (!keyword.value.match(/[\w]+/)) {
+            return Promise.resolve();
```
I think you can just return nothing here because `MaybePromise` does not force you to create a Promise.

gfontorbe replied:
Changed

spoenemann reviewed (approved):
Thanks!
